### DISCUSSION TITLE
Execute timer in sync with system clock

### DISCUSSION THREAD

**Discussion by mashupmark** [SmingHub/Sming#2927] [Q&A] [answered]
Hi, I'm currently trying to use Sming for building a clock. I'm using the `NtpClient` to automatically update the system clock and would now like to execute some logic on every 5th minute.
I already looked at the SystemClock_NTP example and through the docs but couldn't find an easy way to do this.
I'm pretty new to Sming so any help would be appreciated :)

**Comment by mashupmark** [comment]
Ok, I was able to solve this myself. I basically build a "rookie" crontab by having a timer run every second and comparing the current date time inside its callback (1s is precise enough for my use case). E.g.:

```cpp
auto dt = DateTime(SystemClock.now());
// Only continue if it's the 5th minute of the current hour
if (dt.Minute % 5 != 0 && dt.Second < 1)
    return;
```

If there is some built in utility or library I overlooked, feel free to let me know though :)

**Comment by mikee47** [comment, marked as answer]
It's risky checking the clock since we cannot guarantee the time is monotonic: NTP updates may cause forward or backward jumps. It's also not efficient using DateTime like this.

I'd recommend just using a `Timer` configured to fire after the desired interval. Something like this:

```
const unsigned EVENT_INTERVAL_SECS = 60 * 5;
Timer actionTimer;
time_t nextEventTime;

void scheduleNextEvent();

void handleEvent()
{
    time_t now = SystemClock.now(eTZ_UTC);
    scheduleNextEvent();
}

void scheduleNextEvent()
{
    // Base next event on current time to compensate for clock drift
    time_t now = SystemClock.now(eTZ_UTC);
    unsigned secondsSinceLastEvent = now % EVENT_INTERVAL_SECS;
    unsigned secondsUntilNextEvent = EVENT_INTERVAL_SECS - secondsSinceLastEvent;

    // First call we could be very close to the next event time
    if(secondsSinceLastEvent < 5) {
        secondsUntilNextEvent += EVENT_INTERVAL_SECS;
    }

    // Might be useful to take a note of the next due time. Not necessary though.
    nextEventTime = now + secondsUntilNextEvent;

    // Start timer for next event
    actionTimer.initializeMs<1000*secondsUntilNextEvent>(handleEvent);
    actionTimer.startOnce();
}
```

We probably shouldn't schedule the first event until the clock's been set, so let's do that in the clock change handler:

```
void onNtpReceive(NtpClient& client, time_t timestamp)
{
    SystemClock.setTime(timestamp, eTZ_UTC);

    if(nextEventTime == 0) {
        scheduleNextEvent();
    }
}
```
